# Homework 4

## Question 1: 

Unlike the trajectory of perceptrons and expert systems, I believe that deep neural networks have a promising future. In comparison, neural networks have a stronger foundation based on statistical methods and large amounts of data which allow them to tackle complex, not black-and-white situations. Perceptrons and expert systems have been looked down upon because they had very little prospect in solving real-world problems.

Neural networks are capable of solving problems efficiently and quickly. Even when neural networks give partial or less-accurate solutions, those solutions are often still extremely. At work, I’m currently working with Uber’s open source, Ludwig, and it is able to solve real-world problems. For example, it is used to perform Machine Language translations, categorize  (positive/negative) emotions in language, as well as classify and caption images. 

But despite all of the progress and promise neural nets have lend to AI, it’s important to not exxagerate neural networks’ ability. I do not believe that nueral networks will be able to transcend humanity or be able to truly learn and discern like humans. In fact, I believe that neural networks are dependent on humans. A neural network depends on humans to define its goals (even in unsupervised learning- the goals are often driven by the data or are simple goals)  and to feed it data. Therefore, neural networks, I believe can develop to become even more efficient,  faster, and more accurate in not only producing solutions but also handling larger volumes of data. But it won’t be able to transcend human cognition and will eventually meet its limitations. 

Sources: 

https://towardsdatascience.com/limitations-of-deep-learning-in-ai-research-5eed166a4205  
https://arxiv.org/ftp/arxiv/papers/1801/1801.00631.pdf (p. 5-10) 


## Question 2: 

The solution is submitted as an image within the Homework 4 folder

## Question 3:

In [6]:
# this is heavily inspired by kvlinden's keras-cnn example

# import essential libraries and fashion_mnist dataset
from keras import layers
from keras import models
from keras.datasets import fashion_mnist
from keras.utils import to_categorical

In [7]:
model = models.Sequential()

#  3 layer convnet using conv2d and MaxPool
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Add layers to flatten the 2D image and then do a 10-way classification.
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_2 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 64)                36928     
__________

In [8]:

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

#Describe the training and testing set:

print('train_image count:', len(train_images))
print('train_image shape:', train_images.shape)
print('train_image type:', train_images.dtype)
print('train_labels type:', train_labels.dtype)

print('\ntest_image count:', len(test_images))
print('test_image shape:', test_images.shape)

train_image count: 60000
train_image shape: (60000, 28, 28)
train_image type: uint8
train_labels type: uint8

test_image count: 10000
test_image shape: (10000, 28, 28)


In [9]:
train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [10]:
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=6, batch_size=64)
model.evaluate(test_images, test_labels)

Instructions for updating:
Use tf.cast instead.
Epoch 1/6
60000/60000 [==============================] - 32s 539us/step - loss: 0.5243 - acc: 0.8052
Epoch 2/6
60000/60000 [==============================] - 32s 532us/step - loss: 0.3239 - acc: 0.8822
Epoch 3/6
60000/60000 [==============================] - 31s 525us/step - loss: 0.2783 - acc: 0.8987
Epoch 4/6
60000/60000 [==============================] - 31s 523us/step - loss: 0.2472 - acc: 0.9089
Epoch 5/6
60000/60000 [==============================] - 31s 522us/step - loss: 0.2244 - acc: 0.9178
Epoch 6/6
10000/10000 [==============================] - 2s 193us/step


[0.26510446031689644, 0.9042]

This network, of 3 layers, gives me an accuracy of 90.42%